File: app/src/main/java/com/mv2studio/myride/connection/BluetoothManager.java
package com.mv2studio.myride.connection;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;

import com.mv2studio.myride.utils.Log;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.UUID;

/**
 * Created by matej on 30.10.16.
 */

public class BluetoothManager {

    private static final UUID MY_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

    public static BluetoothSocket connect(BluetoothDevice device) {
        if (device == null) return null;

        BluetoothSocket socket = null;

        Log.d("Connecting to BT device.");
        try {
            socket = device.createInsecureRfcommSocketToServiceRecord(MY_UUID);
            socket.connect();
            Log.d("Wooha! Looks like connection works!");
        } catch (IOException ioe) {
            Log.e("Could not connect to remote device. Trying some workaround.", ioe);
            Class<?> clazz = socket.getRemoteDevice().getClass();
            Class<?>[] paramTypes = new Class<?>[]{Integer.TYPE};

            try {
                Method m = clazz.getMethod("createRfcommSocket", paramTypes);
                Object[] params = new Object[]{Integer.valueOf(1)};
                socket = (BluetoothSocket) m.invoke(socket.getRemoteDevice(), params);
                // If this will not work try adding Thread.sleep(500);
                socket.connect();
                Log.d("*sweating* Wooha! Fallback BT connection worked.");
            } catch (Exception e) {
                Log.e("Well, connection does not work. Don't freak out, but I really have no idea what to do now...", e);
            }
        }

        return socket;
    }

}

File: app/src/main/kotlin/com/mv2studio/myride/connection/BluetoothManager.kt
package com.mv2studio.myride.connection

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import com.mv2studio.myride.extensions.d
import com.mv2studio.myride.extensions.e
import java.io.IOException
import java.util.*

/**
 * Created by matej on 15/11/2016.
 */
class BluetoothManager {

    companion object {
        private val MY_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")

        fun connect(device: BluetoothDevice): BluetoothSocket? {
            var socket: BluetoothSocket? = null

            d("Connecting to BT device.")
            try {
                socket = device.createInsecureRfcommSocketToServiceRecord(MY_UUID)
                socket.connect()
            } catch (ioe: IOException) {
                e("Could not connect to remote device. Trying some workaround.", ioe)
                val clazz = socket?.remoteDevice?.javaClass

                try {
                    val m = clazz?.getMethod("createRfcommSocket", Integer.TYPE)
                    socket = m?.invoke(socket?.remoteDevice, 1) as? BluetoothSocket?

                    //TODO If this will not work try adding Thread.sleep(500);
                    socket?.connect()
                    d("*sweating* Wooha! Fallback BT connection worked.")
                } catch (ex: Exception) {
                    e("Well, connection does not work. Don't freak out, but I really have no idea what to do now...", ex)
                }
            }

            return socket
        }
    }

}